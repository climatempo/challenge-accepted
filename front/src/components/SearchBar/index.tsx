import {
  Form,
  Input,
  SearchButton,
  SearchIcon,
  Sugestions,
  Sugestion,
  SugestionButton,
} from "./styles";
import { Props } from "./types";

function SearchBar({
  onSubmit,
  onChange,
  onFocus,
  onBlur,
  handleRouterPush,
  displaySugestions,
  sugestions,
  value,
  isFocused,
}: Props) {
  return (
    <Form onSubmit={onSubmit} isFocused={isFocused}>
      <Input
        type="search"
        placeholder="VitÃ³ria - ES"
        onFocus={onFocus}
        debounceTimeout={500}
        onBlur={onBlur}
        onChange={onChange}
        isFocused={isFocused}
        value={value}
      />
      <SearchButton type="submit">
        <SearchIcon src="/icons/search.png" isFocused={isFocused} />
      </SearchButton>
      {displaySugestions && (
        <Sugestions>
          {sugestions.map(({ id, name, state }) => (
            <Sugestion key={id}>
              <SugestionButton
                onClick={handleRouterPush(id)}
              >{`${name} - ${state}`}</SugestionButton>
            </Sugestion>
          ))}
        </Sugestions>
      )}
    </Form>
  );
}

export default SearchBar;
