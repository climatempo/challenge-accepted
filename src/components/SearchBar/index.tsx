import Wrapper, { Form, Input, SearchButton, SearchIcon } from "./styles";
import { Props } from "./types";

function SearchBar({ onSubmit }: Props) {
  return (
    <Form onSubmit={onSubmit}>
      <Input />
      <SearchButton type="submit">
        <SearchIcon src="/icons/search.png" />
      </SearchButton>
    </Form>
  );
}

export default SearchBar;
