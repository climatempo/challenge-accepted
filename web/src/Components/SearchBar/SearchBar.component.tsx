import React, { FormEvent, useState } from 'react';
import { Autocomplete, Box, TextField } from '@mui/material';
import { useAppDispatch, useAppSelector, useServerAddress } from '../../hooks';
import { DetailedLocale, setLocale } from '../../Reducers/Locale.reducer';
import { useSnackbar } from 'notistack';

function SearchBar() {
    const [options, setOptions] = useState<DetailedLocale[]>([]);
    const pageLocale = useAppSelector(state => state.locale.value);
    const { enqueueSnackbar: notify } = useSnackbar();
    const dispatch = useAppDispatch();
    const { search } = useServerAddress();
    let fetchBlock = false;

    function fetchSearch(event: FormEvent) {
        if(fetchBlock)
            return;

        const input = (event.target as HTMLInputElement).value.trim();
        if(input.length < 3 || input.length > 100)
            return;

        fetch(search(input))
            .then(response => response.json())
            .then(setOptions)
            .catch(err => {
                fetchBlock = true;
                console.error(err);
                notify('Alguma coisa deu errado.', { variant: 'error'});

                setTimeout(() => fetchBlock = false, 1000);
            });
    }

    function setPageLocale(event: FormEvent) {
        if(fetchBlock)
            return;

        const selected = (event.target as HTMLDListElement).getAttribute('data-locale');
        if(!selected)
            return;
        const locale = JSON.parse(selected) as DetailedLocale;
        if(pageLocale && pageLocale.idlocale === locale.idlocale)
            return;

        dispatch(setLocale(locale));
    }

    function onKeyPress(event: React.KeyboardEvent<HTMLDivElement>) {
        if(event.key !== 'Enter')
            return;
        setTimeout(() => {
            const selected = (event.target as HTMLInputElement).value;
            console.log(selected);
            const locale = options.filter(
                locale => locale.city + ', ' + locale.uf.toUpperCase() === selected
            )[0];
            if(!locale)
                return;
            if(pageLocale && pageLocale.idlocale === locale.idlocale)
                return;
            dispatch(setLocale(locale));
        }, 50);
    }

    return <Autocomplete
        options={options}
        onInput={fetchSearch}
        onChange={setPageLocale}
        onKeyDown={onKeyPress}
        fullWidth
        inputMode="text"
        getOptionLabel={locale => locale ? locale.city + ', ' + locale.uf.toUpperCase() : ''}
        isOptionEqualToValue={(option, value) => option.idlocale === value.idlocale}
        renderOption={(props, locale) => (
            <Box component="li"
                 {...props}
                 data-locale={JSON.stringify(locale)}
                 key={locale.idlocale}>
                { locale.city + ', ' + locale.uf.toUpperCase() }
            </Box>
        )}
        renderInput={(params) => (
            <TextField
                {...params}
                hiddenLabel
                variant="outlined"
                label="Buscar cidade"
                fullWidth
                type="search"
                InputProps={{
                    ...params.InputProps,
                    type: 'search',
                }}
            />
        )}
    />;

}

export default SearchBar;