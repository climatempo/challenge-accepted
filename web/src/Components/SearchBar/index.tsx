import Button from './SearchBar.component';

export default Button;