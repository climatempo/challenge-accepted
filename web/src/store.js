import Vue from "vue";
import Vuex from "vuex";
import localeStore from "./store/locale";
import weatherStore from "./store/weather";

Vue.use(Vuex);

export default new Vuex.Store({
  state: {},
  mutations: {},
  actions: {},
  modules: [
      localeStore,
      weatherStore
  ]
});
