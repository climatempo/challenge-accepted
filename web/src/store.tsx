import { configureStore } from '@reduxjs/toolkit';
import themeReducer from './Reducers/Theme.reducer';
import temperatureReducer from './Reducers/Temperature.reducer';
import precipitationReducer from './Reducers/Precipitation.reducer';

const store = configureStore({
    reducer: {
        theme: themeReducer,
        temperature: temperatureReducer,
        precipitation: precipitationReducer,
    }
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export default store;