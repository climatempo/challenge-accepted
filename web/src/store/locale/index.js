import mutations from './mutations'
import actions from './actions'
import getters from './getters'

/**
 * Store de localizaçeõs
 *
 * @type mixins
 */
const localeStore = {
    state: {
        locales: []
    },
    mutations: mutations,
    actions: actions,
    getters: getters
};

export default localeStore
