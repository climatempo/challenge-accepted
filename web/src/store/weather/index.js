import mutations from './mutations'
import actions from './actions'
import getters from './getters'

/**
 * Store de previs√µes
 *
 * @type mixins
 */
const weatherStore = {
    state: {
        weather: []
    },
    mutations: mutations,
    actions: actions,
    getters: getters
};

export default weatherStore
