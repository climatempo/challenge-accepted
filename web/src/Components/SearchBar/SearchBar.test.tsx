import renderer from 'react-test-renderer';
import '../../Tests/watchMedia.mock';
import store from '../../store';
import { Provider } from 'react-redux';
import { SnackbarProvider } from 'notistack';
import SearchBar from './index';

test('Search Bar component', () => {
    const bar = renderer.create(
      <Provider store={store}>
          <SnackbarProvider maxSnack={4}>
              <SearchBar/>
          </SnackbarProvider>
      </Provider>
    );
    expect(bar.toJSON()).toMatchSnapshot();
});