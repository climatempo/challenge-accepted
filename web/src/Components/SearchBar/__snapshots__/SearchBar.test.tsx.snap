// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Bar component 1`] = `
<div
  aria-owns={null}
  className="MuiAutocomplete-root MuiAutocomplete-fullWidth MuiAutocomplete-hasPopupIcon css-qm9mjt-MuiAutocomplete-root"
  inputMode="text"
  onClick={[Function]}
  onInput={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={false}
      htmlFor=":r0:"
      id=":r0:-label"
    >
      Buscar cidade
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiInputBase-hiddenLabel MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-expanded={false}
        aria-invalid={false}
        autoCapitalize="none"
        autoComplete="off"
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputTypeSearch MuiInputBase-inputHiddenLabel MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-1j5luls-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id=":r0:"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        readOnly={false}
        required={false}
        role="combobox"
        spellCheck="false"
        type="search"
        value=""
      />
      <div
        className="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
      >
        <button
          aria-label="Open"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Open"
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </button>
      </div>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-yjsfm1"
        >
          <span>
            Buscar cidade
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
